### PR TITLE
cli: avoid logging command line errors in more cases

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

"kubectl list" should print a plain text explanation ("Unknown command "list"
for kubectl. ...") without treating that multi-line error as a log message.

The previous heuristic didn't work here because the command itself was not
found. A better one is to check that command execution really started.

This is still not perfect: what if a command hasn't started logging yet or
never uses logging for its output and then returns an error, for example in
`kubectl kustomize`? To handle that case, `kubectl` now always deals
with error printing itself.

#### Which issue(s) this PR fixes:
Fixes #https://github.com/kubernetes/kubernetes/issues/107012

#### Does this PR introduce a user-facing change?
```release-note
Fix 1.23 regression: Some command line errors (for example, "kubectl list" -> "unknown command") were printed as log message with escaped line breaks instead of a multi-line plain text, which made the error harder to read.
```

### CODE REVIEW

File: cmd/kubectl/kubectl.go [hunk 1]
```diff
@@ -17,17 +17,18 @@ limitations under the License.
...
 
 	// Import to initialize client auth plugins.
 	_ "k8s.io/client-go/plugin/pkg/client/auth"
 )
 
 func main() {
 	command := cmd.NewDefaultKubectlCommand()
-	code := cli.Run(command)
-	os.Exit(code)
+	if err := cli.RunNoOutput(command); err != nil {
```
Comment (xyz-li) [inline cmd/kubectl/kubectl.go:30]: RunNoErrOutput

Comment (pohly) [reply cmd/kubectl/kubectl.go:30]: Darn. Well, now you know what my original name was before a last-minute change of mind...

**Comment by pohly** [comment]
/retest

@xyz-li: PR should be okay now, there was just one flake in pull-kubernetes-integration.

**Comment by pohly** [comment]
The `deprecated_api` test-cmd case seems to be sensitive to the output format. It's failing consistently at the moment:
```
+++ [0106 03:10:59] Testing deprecated APIs
Successful
message:Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
No resources found
has:PodSecurityPolicy is deprecated
Successful
message:Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
No resources found
error: 1 warning received
has:PodSecurityPolicy is deprecated
FAIL!
message:Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
No resources found
error: 1 warning received
has not:err="1 warning received"
424 /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../test/cmd/get.sh
```

For comparison, on master the output is:
```
+++ Running case: test-cmd.run_deprecated_api_tests 
+++ working dir: /nvme/gopath/src/k8s.io/kubernetes
+++ command: run_deprecated_api_tests
+++ [0110 16:56:01] Creating namespace namespace-1641830161-21779
namespace/namespace-1641830161-21779 created
Context "test" modified.
+++ [0110 16:56:01] Testing deprecated APIs
Successful
message:Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
No resources found
has:PodSecurityPolicy is deprecated
Successful
message:Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
No resources found
E0110 16:56:01.987555  202060 run.go:120] "command failed" err="1 warning received"
has:PodSecurityPolicy is deprecated
Successful
message:Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
No resources found
E0110 16:56:01.987555  202060 run.go:120] "command failed" err="1 warning received"
has:err="1 warning received"
+++ exit code: 0
warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
No resources found
warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
No resources found
+++ [0110 16:56:02] TESTS PASSED
```

**Comment by pohly** [comment]
I think ending with `error: 1 warning received` is the cleaner output and therefore I have adapted the test case to match that.

**Review by eddiezane — COMMENTED** [review]
Can you please add an integration test to catch a regression here?

**Comment by pohly** [comment]
> Can you please add an integration test to catch a regression here?

@eddiezane: With integration test, do you mean something like https://github.com/kubernetes/kubernetes/blob/a0dfd958d5c5e8c2bd5fd30e1d2e9031e9aba1b6/test/cmd/get.sh#L411-L428?

**Comment by pohly** [comment]
I've added such a test for "kubectl list" and "kubectl get pods/no-such-pods". See latest commit message for details.

There's nothing special about "kubectl get pods/no-such-pods", but it can be tested because the output should be deterministic. More test cases would be even better, to catch regressions *before* they occur... :sweat_smile:

File: cmd/kubectl/kubectl.go [hunk 2]
```diff
@@ -17,17 +17,18 @@ limitations under the License.
 package main
 
 import (
-	"os"
-
 	"k8s.io/component-base/cli"
 	"k8s.io/kubectl/pkg/cmd"
+	"k8s.io/kubectl/pkg/cmd/util"
```
Comment (soltysh) [inline cmd/kubectl/kubectl.go:22]: Nit: `cmdutil` is the preferred alias for these to make it more explicit.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Pull request merged**